### PR TITLE
Added validation check to impose limits on stored, free-form strings 

### PR DESCRIPTION
**Pull request by evalsocket**
- Added client side validation in flyteadmin 
- Added 255 char limit in all string field 

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/lyft/flyte/issues/336
https://github.com/lyft/flyte/issues/896

## Follow-up issue
_NA_
OR
_https://github.com/lyft/flyte/issues/<number>_

### CODE REVIEW

File: pkg/repositories/models/project.go [hunk 1]
```diff
@@ -3,7 +3,7 @@ package models
 type Project struct {
 	BaseModel
 	Identifier  string `gorm:"primary_key"`
-	Name        string // Human-readable name, not a unique identifier.
+	Name        string `valid:"length(3|50)"` // Human-readable name, not a unique identifier.
```
Comment (katrogan) [inline pkg/repositories/models/project.go:6]: how does this work? does this require a db migration?

**Comment by evalsocket** [comment]
@kumare3 I don't need to list all string fields...you can comment on my PR and I will update the changes 
cc @katrogan @EngHabu

**Comment by kumare3** [comment]
Aah ya sure

**Comment by evalsocket** [comment]
@kumare3 Any update on this ?

**Comment by kumare3** [comment]
Will look tomorrow

File: pkg/repositories/models/execution.go [hunk 2]
```diff
@@ -37,7 +37,7 @@ type Execution struct {
 	ExecutionEvents    []ExecutionEvent
 	// In the case of an aborted execution this string may be non-empty.
 	// It should be ignored for any other value of phase other than aborted.
-	AbortCause string
+	AbortCause string `valid:"length(3|50)"`
```
Comment (kumare3) [inline pkg/repositories/models/execution.go:40]: this should be 255 chars

File: pkg/repositories/models/launch_plan.go [hunk 3]
```diff
@@ -2,10 +2,10 @@ package models
 
 // Launch plan primary key
 type LaunchPlanKey struct {
-	Project string `gorm:"primary_key;index:lp_project_domain_name_idx,lp_project_domain_idx"`
-	Domain  string `gorm:"primary_key;index:lp_project_domain_name_idx,lp_project_domain_idx"`
-	Name    string `gorm:"primary_key;index:lp_project_domain_name_idx"`
-	Version string `gorm:"primary_key"`
+	Project string `gorm:"primary_key;index:lp_project_domain_name_idx,lp_project_domain_idx" valid:"length(3|50)"`
+	Domain  string `gorm:"primary_key;index:lp_project_domain_name_idx,lp_project_domain_idx" valid:"length(3|50)"`
+	Name    string `gorm:"primary_key;index:lp_project_domain_name_idx" valid:"length(3|50)"`
```
Comment (kumare3) [inline pkg/repositories/models/launch_plan.go:7]: i think this can be larger right? @katrogan

File: pkg/repositories/models/named_entity.go [hunk 4]
```diff
@@ -6,10 +6,10 @@ import (
...
 // NamedEntityMetadata primary key
 type NamedEntityMetadataKey struct {
-	ResourceType core.ResourceType `gorm:"primary_key;index:named_entity_metadata_type_project_domain_name_idx"`
-	Project      string            `gorm:"primary_key;index:named_entity_metadata_type_project_domain_name_idx"`
-	Domain       string            `gorm:"primary_key;index:named_entity_metadata_type_project_domain_name_idx"`
-	Name         string            `gorm:"primary_key;index:named_entity_metadata_type_project_domain_name_idx"`
+	ResourceType core.ResourceType `gorm:"primary_key;index:named_entity_metadata_type_project_domain_name_idx" valid:"length(3|50)"`
+	Project      string            `gorm:"primary_key;index:named_entity_metadata_type_project_domain_name_idx" valid:"length(3|50)"`
+	Domain       string            `gorm:"primary_key;index:named_entity_metadata_type_project_domain_name_idx" valid:"length(3|50)"`
+	Name         string            `gorm:"primary_key;index:named_entity_metadata_type_project_domain_name_idx" valid:"length(3|50)"`
```
Comment (kumare3) [inline pkg/repositories/models/named_entity.go:12]: same here

File: pkg/repositories/models/task.go [hunk 5]
```diff
@@ -5,10 +5,10 @@ package models
 
 // Task primary key
 type TaskKey struct {
-	Project string `gorm:"primary_key;index:task_project_domain_name_idx,task_project_domain_idx"`
-	Domain  string `gorm:"primary_key;index:task_project_domain_name_idx,task_project_domain_idx"`
-	Name    string `gorm:"primary_key;index:task_project_domain_name_idx"`
-	Version string `gorm:"primary_key"`
+	Project string `gorm:"primary_key;index:task_project_domain_name_idx,task_project_domain_idx" valid:"length(3|50)"`
+	Domain  string `gorm:"primary_key;index:task_project_domain_name_idx,task_project_domain_idx" valid:"length(3|50)"`
+	Name    string `gorm:"primary_key;index:task_project_domain_name_idx" valid:"length(3|50)"`
```
Comment (kumare3) [inline pkg/repositories/models/task.go:10]: same here

File: pkg/repositories/models/task_execution.go [hunk 6]
```diff
@@ -20,9 +20,9 @@ type TaskExecutionKey struct {
 type TaskExecution struct {
 	BaseModel
 	TaskExecutionKey
-	Phase        string
+	Phase        string `valid:"length(3|50)"`
 	PhaseVersion uint32
-	InputURI     string
+	InputURI     string `valid:"length(3|50)"`
```
Comment (kumare3) [inline pkg/repositories/models/task_execution.go:25]: this can be pretty large. 255 chars

File: pkg/repositories/models/workflow.go [hunk 7]
```diff
@@ -13,7 +13,7 @@ type Workflow struct {
 	BaseModel
 	WorkflowKey
 	TypedInterface          []byte
-	RemoteClosureIdentifier string `gorm:"not null"`
+	RemoteClosureIdentifier string `gorm:"not null" valid:"length(3|50)"`
```
Comment (kumare3) [inline pkg/repositories/models/workflow.go:16]: i dont know what this string it - @katrogan

**Comment by kumare3** [comment]
@evalsocket this can be a very dangerous change. I would need reviews from @anandswaminathan / @katrogan / @kanterov

File: pkg/repositories/models/workflow.go (hunk 7 above)
Comment (katrogan) [reply pkg/repositories/models/workflow.go:16]: this is the path to where on s3 (or whichever cloud blobstore) we upload the workflow closure

File: tests/launch_plan_test.go [hunk 8]
```diff
@@ -237,6 +237,7 @@ func TestEnableDisableLaunchPlan(t *testing.T) {
 	createWorkflowReq := getWorkflowCreateRequest()
 
 	_, err := client.CreateWorkflow(ctx, &createWorkflowReq)
+	fmt.Println(err)
```
Comment (kanterov) [inline tests/launch_plan_test.go:240]: Looks unintentionally changed

**Comment by kanterov** [comment]
Does `length(3|50)` means that there is a lower bound of 3? What is the reason for that?

**Comment by evalsocket** [comment]
@kanterov 3 is lower bound and we can remove it If we don't want it.

**Comment by evalsocket** [comment]
@katrogan @kumare3 @anandswaminathan  @kanterov  need your review?

File: pkg/repositories/models/execution.go [hunk 9]
```diff
@@ -11,9 +11,9 @@ import (
 
 // Execution primary key
 type ExecutionKey struct {
-	Project string `gorm:"primary_key;column:execution_project"`
-	Domain  string `gorm:"primary_key;column:execution_domain"`
-	Name    string `gorm:"primary_key;column:execution_name"`
+	Project string `gorm:"primary_key;column:execution_project" valid:"length(3|255)"`
```
Comment (katrogan) [inline pkg/repositories/models/execution.go:14]: how did we settle on 3 as a min char limit?

File: pkg/repositories/models/execution.go [hunk 10]
```diff
@@ -37,7 +37,7 @@ type Execution struct {
 	ExecutionEvents    []ExecutionEvent
 	// In the case of an aborted execution this string may be non-empty.
 	// It should be ignored for any other value of phase other than aborted.
-	AbortCause string
+	AbortCause string `valid:"length(0|50)"`
```
Comment (anandswaminathan) [inline pkg/repositories/models/execution.go:40]: 50 seems low ? Have it at 255 as well ?

File: pkg/repositories/models/named_entity.go [hunk 11]
```diff
@@ -30,9 +30,9 @@ type NamedEntityMetadata struct {
 // fields here should match the ones in NamedEntityMetadataKey.
 type NamedEntityKey struct {
 	ResourceType core.ResourceType
-	Project      string
-	Domain       string
-	Name         string
+	Project      string `valid:"length(0|50)"`
+	Domain       string `valid:"length(0|50)"`
+	Name         string `valid:"length(0|50)"`
```
Comment (anandswaminathan) [inline pkg/repositories/models/named_entity.go:35]: Name is 255 is some places, and 50 in others? Can we have it uniform at 255 ?

File: pkg/repositories/models/resource.go [hunk 12]
```diff
@@ -18,11 +18,11 @@ type Resource struct {
 	CreatedAt    time.Time
 	UpdatedAt    time.Time
 	DeletedAt    *time.Time `sql:"index"`
-	Project      string     `gorm:"unique_index:resource_idx"`
-	Domain       string     `gorm:"unique_index:resource_idx"`
-	Workflow     string     `gorm:"unique_index:resource_idx"`
-	LaunchPlan   string     `gorm:"unique_index:resource_idx"`
-	ResourceType string     `gorm:"unique_index:resource_idx"`
+	Project      string     `gorm:"unique_index:resource_idx" valid:"length(0|50)"`
```
Comment (anandswaminathan) [inline pkg/repositories/models/resource.go:21]: Same as above comment. Why have 255 at some places, and 50 here?

**Comment by evalsocket** [comment]
@anandswaminathan  As per your comment. I added an upper limit to 255 char as a uniform value rather than 50 char.  
CC: @kumare3 @kanterov

File: pkg/repositories/models/node_execution.go [hunk 13]
```diff
@@ -9,7 +9,7 @@ import (
 
 type NodeExecutionKey struct {
 	ExecutionKey
-	NodeID string `gorm:"primary_key;index"`
+	NodeID string `gorm:"primary_key;index" valid:"length(0|50)"`
```
[the remark this answers is not in the document]
Comment (evalsocket) [reply pkg/repositories/models/node_execution.go]: @EngHabu Any comment ?

[the remark this answers is not in the document]
Comment (EngHabu) [reply pkg/repositories/models/node_execution.go]: This is a valid concern, migration will either fail or silently ignore the constraint... do you mind testing that @evalsocket ? I tried to quickly google it but couldn't find information, so you need to create a workflow with a node that has a long name ([example](https://github.com/flyteorg/flytekit/blob/e373b504fad7d7d323d94762a2c91b41b1043b6d/tests/flytekit/unit/core/test_node_creation.py#L91)) and then attempt to deploy the DB migration and see..

[the remark this answers is not in the document]
Comment (evalsocket) [reply pkg/repositories/models/node_execution.go]: ok i will do it tomorrow

File: go.sum [hunk 14]
```diff
@@ -170,6 +170,7 @@ github.com/aws/aws-sdk-go-v2/service/sts v1.0.0/go.mod h1:5f+cELGATgill5Pu3/vK3E
 github.com/aws/aws-xray-sdk-go v0.9.4/go.mod h1:XtMKdBQfpVut+tJEwI7+dJFRxxRdxHDyVNp2tHXRq04=
 github.com/aws/smithy-go v1.0.0/go.mod h1:EzMw8dbp/YJL4A5/sbhGddag+NPT7q084agLbB9LgIw=
 github.com/aws/smithy-go v1.1.0/go.mod h1:EzMw8dbp/YJL4A5/sbhGddag+NPT7q084agLbB9LgIw=
+github.com/aymerick/douceur v0.2.0/go.mod h1:wlT5vV2O3h55X9m7iVYN0TBM0NH/MmbLnd30/FjWUq4=
```
Comment (EngHabu) [inline go.sum:173]: mind running `go mod tidy`

File: pkg/repositories/models/node_execution.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (EngHabu) [reply pkg/repositories/models/node_execution.go]: Just to be clear, what did you do to validate? you created this WF, registered it (success), then applied the DB migration (success?)? then tried to register again (client-side failure)?

[the remark this answers is not in the document]
Comment (evalsocket) [reply pkg/repositories/models/node_execution.go]: Test case: @EngHabu 
- Created a workflow with large node name 
- Flytekit trimmed the node name on serialize time and only get 64 char
- manually read and write the serialize workflow and edit the large node name
- Register workflow successfully 
- Migration worked fine 
- Try to launch the workflow but got 500......In logs got db validation error as expected (`to db with err unexpected error type for: AbortCause is the wrong length (should be 1~255 characters); Cluster is the wrong length (should be 1~255 characters); User is the wrong length (should be 1~255 characters)`)

File: go.sum (hunk 14 above)
Comment (evalsocket) [reply go.sum:173]: It's already done

File: go.sum [hunk 15]
```diff
@@ -1203,6 +1211,12 @@ github.com/prometheus/procfs v0.3.0/go.mod h1:lV6e/gmhEcM9IjHGsFOCxxuZ+z1YqCvr4O
 github.com/prometheus/procfs v0.6.0 h1:mxy4L2jP6qMonqmq+aTtOx1ifVWUgG/TAmntgbh3xv4=
 github.com/prometheus/procfs v0.6.0/go.mod h1:cz+aTbrPOrUb4q7XlbU9ygM+/jj0fzG6c1xBZuNvfVA=
 github.com/prometheus/tsdb v0.7.1/go.mod h1:qhTCs0VvXwvX/y3TZrWD7rabWM+ijKTux40TwIPHuXU=
+github.com/qor/qor v0.0.0-20200729071734-d587cffbbb93 h1:kxseRCbX7NxSvKYSO1rXQCALzl3YLZvooxhhQdA23fU=
```
Comment (EngHabu) [inline go.sum:1214]: Please run `go mod tidy`

**Review by EngHabu — APPROVED** [review]

**Review by EngHabu — APPROVED** [review]

**Pull request merged**